streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -436,6 +496,8 @@ private void maybeSetError() {
             }
 
             if (setState(State.ERROR)) {
+                metrics.close();
```

Comment:
If the client is in error we should close these threads, for the same reason we are closing them for this kip. So might as well close them now.

This is also ensures they are closed on all clients after a shutdown

Reply:
Actually, now that I have a more in-depth picture of what is going on, I disagree. I think we should leave these threads running until users actually call `KafkaStreams.close`, since the could alternatively add more threads (via KIP-633) to transition back from ERROR state into RUNNING, at which point, we'll be sorry that we killed these threads, right?

Reply:
Maybe when we transition out of error we should restart them. I don't think the metrics thread really matters, but I closed the the state cleaner bc @cadonna thought it might cause state loss we would want to avoid

Reply:
It certainly might. I'm just wary of how far into the uncanny valley we're going here. Streams is going to be put into a state that's very similar to the one that `close()` produces, but not identical. What will then happen when they _do_ call close?

OTOH, we could instead change direction on the "error-vs-shutdown" debase and just make all these methods call `close(ZERO)` instead. Then, the _real_ close method will be invoked, and Streams will go through a well-known transition through `PENDING_SHUTDOWN` to `NOT_RUNNING`.

It would then be a problem for a later date (after KIP-663) if someone wanted to request that instead the app should stop all running threads so they can manually call "addThread" later.

Reply:
We did have it this way in the kip. If we stick to this for now I think that we can clear this up easily when we decide what we want to do with the sates in general in when we take care of the discussion in KIP-663